Accept attributes design & remove spec note (#845)
unicode-org/message-format-wg: spec/formatting.md

as long as the final _formatting_ result is made available to users
and the observable behavior of the formatter matches that described here.

> _Attributes_ MUST NOT affect the processing or output of a _message_.

Reviewer: This might be too broadly stated, since the purpose of attributes might be the "processing" of a message (as, for example, in tools). 

If we wanted to prevent attributes from having an effect, we'd be better off to say that implementations must not/should not expose attribute values to functions (if functions can't see the attributes, they can't do anything with them)?

Perhaps:

```suggestion
_Attributes_ have no effect on selection or formatting in a _message_.
```

... and perhaps adding:

> Implementations MUST NOT pass _attributes_ to _functions_.

Author: I think we need a slightly wider ban than attributes' visibility to functions, to prohibit e.g. the formatted parts representation of a message from reflecting attribute values.

Would this work?

```suggestion
_Attributes_ MUST NOT have any effect on the formatted output of a _message_,
or be made available to function implementations.
```

Reviewer: I suggested a no-MUSTard version in part because I don't know all the use cases that an expression attribute might be put to and in part because I don't know how to test it. If an implementation decided to support some magical attributes, how would it be harmful if (for example) they affected the formatted parts? We don't define any expression attributes, so we could leave it up to implementations to decide what they mean. The statement I suggest should guide implementers away from using attributes as some sort of option factory.

Author: > If an implementation decided to support some magical attributes, how would it be harmful if (for example) they affected the formatted parts?

The MUST NOT establishes a promise that when a message is being processed for formatting only, all of the attributes can be trimmed out. This allows for potentially very verbose content to be included in attributes, given that doing so has no impact on the runtime size or parsing requirements.

Without the mustard, it is much more difficult to make this guarantee, as tools can't be certain about what they can and cannot trim out. It would also significantly fuzzify the boundary between "options" and "attributes", if they can have the same effect.

Reviewer 2: So @locale=fr cannot be an attribute, right? Because it would affect the results of formatting.

It sounds like 'attributes' are really something like 'metadata'

Reviewer 2: Ah, sounds like u: options now handle cases like locale settings?

Author: > So @Locale=fr cannot be an attribute, right? Because it would affect the results of formatting.

That's right.

> It sounds like 'attributes' are really something like 'metadata'

Yes. I would be entirely happy with renaming them as metadata, in case there's a sense that that would make it clearer.

> Ah, sounds like u: options now handle cases like locale settings?

Yes, PR #846 introduces a [`u:locale`](https://github.com/unicode-org/message-format-wg/blob/3a3d4abd84d074509eb181865e9fc0f1f596a33d/spec/registry/unicode.md#ulocale) option for just that.
